File: js/1.7.js
const EPSILON = 0.001

const avg = (x, y) => (x + y) / 2

const isGoodEnough = (guess, prevGuess) => Math.abs(guess - prevGuess) < EPSILON

const improve = (guess, x) => avg(x / guess, guess)

const sqrtIter = (guess, x, prevGuess = 0) => isGoodEnough(guess, prevGuess) ? guess : sqrtIter(improve(guess, x), x, guess)

const sqrt = x => sqrtIter(1, x)

console.log(sqrt(0.2))

File: 1.7.clj
; improve good-enough? procedure
; instead of using fixed precision, compare previous and current guess and if they start to differ slighly, stop
; ✅ the sqrt will work better with very big and very small numbers
(defn abs [x] (if (< x 0) (* -1 x) x))

(defn good-enough? [guess prev-guess] (< (abs (- prev-guess guess)) 0.001))

(defn average [x y] (/ (+ x y) 2.0))

(defn improve [guess x] (average guess (/ x guess)))

(defn sqrt-iter [guess x prev-guess]
  (if (good-enough? guess prev-guess)
          guess
          (sqrt-iter (improve guess x) x guess)))

(defn sqrt [x] (sqrt-iter 1 x 0))

(println (sqrt 0.2))